docs/tutorials/forum-application/7.vote-extension.md

```diff
@@ -0,0 +1,125 @@
+the same `vote extension.`
+
+When a validator receives a `pre-commit` vote with an attached `vote extension`, it calls `VerifyVoteExtension` to
+validate the `vote extension.`
+
+If the vote extension is successfully verified, the `pre-commit` vote is included in the tally. If validation fails,
+the entire `pre-commit` message is ignored.
+
+**Following is the blurb of code for the `VerifyVoteExtension` function:**
+
+```go
+// VerifyVoteExtension verifies the vote extensions and ensure they include the curse words
+// It will not be called for extensions generated by this validator
+func (app *ForumApp) VerifyVoteExtension(_ context.Context, req *abci.VerifyVoteExtensionRequest) (*abci.VerifyVoteExtensionResponse, error) {
+	fmt.Println("Executing Application VerifyVoteExtension")
+
+	if _, ok := app.valAddrToPubKeyMap[string(req.ValidatorAddress)]; !ok {
+		// we do not have a validator with this address mapped; this should never happen
+		panic(fmt.Errorf("unknown validator"))
+	}
+	curseWords := strings.Split(string(req.VoteExtension), "|")
+	tmpCurseWordMap := make(map[string]struct{})
+	// Verify that we do not have double words and the validator is not trying to cheat us
+	for _, word := range curseWords {
+		tmpCurseWordMap[word] = struct{}{}
+	}
+	if len(tmpCurseWordMap) < len(curseWords) {
+		// Extension repeats words
+		return &abci.VerifyVoteExtensionResponse{Status: abci.VERIFY_VOTE_EXTENSION_STATUS_REJECT}, nil
+	}
+	return &abci.VerifyVoteExtensionResponse{Status: abci.VERIFY_VOTE_EXTENSION_STATUS_ACCEPT}, nil
+}
+```
+
+**Explanation of code:**
+
+`VerifyVoteExtension` function takes two parameters: a `context.Context` and a pointer to an `VerifyVoteExtensionRequest` object.
+It returns a pointer to an `VerifyVoteExtensionResponse` object and an error.
+
+The implementation checks if the validator address provided in the request (`req.ValidatorAddress`) is mapped to a public
+key in the `app.valAddrToPubKeyMap.` If the validator address is not found in the map, it panics with an error indicating
+an "unknown validator". This check ensures that the validator making the request is recognized by the application.
+
+The method splits the `VoteExtension` field of the request (`req.VoteExtension`) into individual words using the
+`strings.Split` function. The separator used is the pipe character (|). The resulting words are stored in the `curseWords slice.`
+
+The implementation creates a temporary map called `tmpCurseWordMap` to verify that there are no duplicate words in the
+`curseWords` slice and to check if the validator is trying to cheat by including the same word multiple times.
+
+The code iterates over each word in the `curseWords` slice and adds it as a key to the `tmpCurseWordMap`. Since maps
+in Go cannot have duplicate keys, this process effectively checks for duplicate words.
+
+If the length of the `tmpCurseWordMap` is less than the length of the `curseWords` slice, it means that there are
```

Comment:
We need to check this in `PrepareProposal` when we read the vote extensions, as there may be some vote extensions that didn't undergo `VerifiyVoteExtension` (those that arrived after decision time)

Reply:
added the logic in PrepareProposal too